### PR TITLE
Fix - Upcoming `arithmetic_side_effects` lints

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
This PR is another preparation for #32961.
`saturating` and `wrapping` versions of `div` and `rem` will no longer be accepted.

#### Summary of Changes
Replaces them with their `checked` version.

### CODE REVIEW

**Review by apfitzge — APPROVED** [review]

**Comment by t-nelson** [comment]
i'd like to review some of the introduced unwraps here before merging

File: perf/src/sigverify.rs [hunk 1]
```diff
@@ -614,13 +614,19 @@ pub fn ed25519_verify(
...
+            .checked_div(total_packet_count)
+            .map(|percent_valid| percent_valid < 90)
+    });
+    match maybe_use_cpu {
+        Some(true) => {
+            ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
+            return;
+        }
+        Some(false) => (),
+        None => return,
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply perf/src/sigverify.rs:629]: yeah i assumed as much but didn't want to introduce a panic on that assumption because we aren't sufficiently leveraging type safety to ensure any (un)intentional breakage of that upstream guarantee gets handled here

[the remark this answers is not in the document]
Comment (apfitzge) [reply perf/src/sigverify.rs:629]: yeah, I think just returning here is better than introducing a panic. I'm guessing we won't hit this case, but not entirely sure about that. Regardless if we hit it or not, it is valid to return immediately if we have no packets to verify.

**Comment by apfitzge** [comment]
> i'd like to review some of the introduced unwraps here before merging

I think the main ones that stood out to me are the ones with the compute budget. Those appear to be defined as constants initialized in `ComputeBudget::new`, but not entirely clear those aren't modified and set to 0.

Other unwraps involving `align_of` or constants should be safe as those cannot or are not 0.

File: perf/src/sigverify.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilya-bobyr) [reply perf/src/sigverify.rs:629]: `warpping_div()` would panic if I try to divide by 0[^1].
So, `unwrap()` would not have added a new panic.

It seems nice to handle division by 0 gracefully.
But the `Option<bool>` and the `then_some()`/`or_else()` chain seems to obscure the actual intent.

How about this:
```rust
    let Some(valid_percentage) = 100usize
        .wrapping_mul(valid_packet_count)
        .checked_div(total_packet_count) else {
        return;
    };
    let use_cpu = valid_percentage < 90 || valid_packet_count < 64;
```

[^1]: Here is a simple example:
    ```rust
    fn main() {
        assert_eq!(100i8.wrapping_div(0), 0);
    }
    ```
    ```
    Compiling playground v0.0.1 (/playground)
        Finished dev [unoptimized + debuginfo] target(s) in 0.41s
         Running `target/debug/playground`
    thread 'main' panicked at 'attempt to divide by zero', /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/num/mod.rs:290:5
    note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59ee0662f2a8eff8bb6a9ed33bc5c566#)
    ```
    
    https://play.rust-lang.org/?gist=59ee0662f2a8eff8bb6a9ed33bc5c566

**Comment by Lichtso** [comment]
I made all the divisions which have a non-constant divisor (those used in CU consumption) to be `.unwrap_or(u64::MAX)`. Meaning that if the divisor becomes `0` the program definitely fails because it consumes all CUs.

**Review by t-nelson — APPROVED** [review]
pushed a few more commits to wrap up `arithmetic_side_effects` compatibility

File: programs/bpf_loader/src/syscalls/mod.rs [hunk 2]
```diff
@@ -763,7 +773,7 @@ declare_syscall!(
                 let cost = compute_budget.mem_op_base_cost.max(
                     compute_budget
                         .sha256_byte_cost
-                        .saturating_mul((val.len() as u64).saturating_div(2)),
+                        .saturating_mul((val.len() as u64).checked_div(2).unwrap()),
```
[the remark this answers is not in the document]
Comment (ilya-bobyr) [reply programs/bpf_loader/src/syscalls/mod.rs]: Does it generate a linter warning?
At `-O` it is compiled to a bit shift:

```rust
pub fn square(num: u64) -> u64 {
    num.checked_div(2).unwrap()
}
```
```asm
example::square:
        mov     rax, rdi
        shr     rax
        ret
```
https://godbolt.org/z/xz8cPjccP

File: perf/src/sigverify.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply perf/src/sigverify.rs]: >  So, unwrap() would not have added a new panic.

it would have introduced a _misleading_ `unwrap()`. unwrapped `None` vs. attempt to divide by zero. oughta be obvious upon inspection, but still

File: perf/src/sigverify.rs [hunk 3]
```diff
@@ -614,13 +614,15 @@ pub fn ed25519_verify(
...
-        return ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
+    let Some(valid_percentage) = 100usize
+        .wrapping_mul(valid_packet_count)
+        .checked_div(total_packet_count) else {
+        return;
+    };
+    let use_cpu = valid_percentage < 90 || valid_packet_count < 64;
+    if use_cpu {
+        ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
+        return;
```
Comment (t-nelson) [inline perf/src/sigverify.rs:625]:
```suggestion
    let maybe_valid_percentage = 100usize
        .wrapping_mul(valid_packet_count)
        .checked_div(total_packet_count);
    let Some(valid_percentage) = maybe_valid_percentage else {
        return;
    };
    if valid_percentage < 90 || valid_packet_count < 64 {
        ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
        return;
```

File: perf/src/sigverify.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilya-bobyr) [reply perf/src/sigverify.rs]: I think, ideally, `unwarp()` should not be used in production code %)
`expect()` is better, as it allows the author to be explicit as to why they think this case should never happen.
With `expect()` the error message could be even better than the one you get from `wrapping_div()`:
```rust
let v = 100usize
    .wrapping_mul(valid_packet_count)
    .checked_div(total_packet_count)
    .expect("`total_packet_count` is non-zero because ...");
```

File: programs/bpf_loader/src/syscalls/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lichtso) [reply programs/bpf_loader/src/syscalls/mod.rs]: I think that happens in LLVM, not in rustc. In other words, rustc is probably unaware.

File: perf/src/perf_libs.rs [hunk 4]
```diff
@@ -140,11 +140,13 @@ fn find_cuda_home(perf_libs_path: &Path) -> Option<PathBuf> {
...
-        path + ":" + &env::var("LD_LIBRARY_PATH").unwrap_or_else(|_| "".to_string());
-    info!("setting ld_library_path to: {:?}", ld_library_path);
-    env::set_var("LD_LIBRARY_PATH", ld_library_path);
+pub fn append_to_ld_library_path(mut ld_library_path: String) {
+    if let Ok(env_value) = env::var("LD_LIBRARY_PATH") {
+        ld_library_path.push(':');
+        ld_library_path.push_str(&env_value);
+        info!("setting ld_library_path to: {:?}", ld_library_path);
+        env::set_var("LD_LIBRARY_PATH", ld_library_path);
+    }
```
Comment (apfitzge) [inline perf/src/perf_libs.rs:149]: this isn't the same. If `LD_LIBRARY_PATH` starts off empty, we will not add to it here.

File: sdk/cargo-build-sbf/src/main.rs [hunk 5]
```diff
@@ -105,10 +106,7 @@ where
             writeln!(out, "{key}=\"{value}\" \\").unwrap();
         }
         write!(out, "{}", program.display()).unwrap();
-        for arg in args.iter() {
-            write!(out, " {}", arg.as_ref().to_str().unwrap_or("?")).unwrap();
-        }
-        writeln!(out).unwrap();
+        writeln!(out, "{}", msg).unwrap();
```
Comment (apfitzge) [inline sdk/cargo-build-sbf/src/main.rs:109]: why this change?
we used to print each arg on a separate line. this will now print tem on the same line separated by spaces, right?

File: perf/src/perf_libs.rs (hunk 4 above)
Comment (t-nelson) [reply perf/src/perf_libs.rs:149]: nice catch! will fix momentarily

File: sdk/cargo-build-sbf/src/main.rs (hunk 5 above)
Comment (t-nelson) [reply sdk/cargo-build-sbf/src/main.rs:109]: where were we getting new lines in the removed code?

**Comment by t-nelson** [comment]
> was the commit that switched unwraps to expects dropped intentionally/

n/m, i see that it was squashed in.  more rebase hell than i was hoping for this early in the day :sweat_smile:

File: perf/src/perf_libs.rs (hunk 4 above)
Comment (t-nelson) [reply perf/src/perf_libs.rs]: fixed in [f3aba37](https://github.com/solana-labs/solana/pull/33000/commits/f3aba3707b9de7674cdbcfd8a7d94802f1fff7e7)

File: sdk/cargo-build-sbf/src/main.rs (hunk 5 above)
Comment (apfitzge) [reply sdk/cargo-build-sbf/src/main.rs:109]: nevermind I am an idiot. this is fine

**Review by apfitzge — APPROVED** [review]
